REF: convert LegacyWallet to typescript

demerzel3 opened the pull request:
Couple notes for the reviewer:
* the `getAddress` method of `LegacyWallet` usually returns `string` but can also return `false` in some cases; adding `false` to the type exposed some issues in the rest of the code that I tried to handle case by case in a sensible manner. in case you don't like my changes or don't want to risk changing code that is working I can rollback those changes and add an ignore comment or similar to make TS happy
* renaming `legacy-wallet.js` to `legacy-wallet.ts` changed the way the file is transpiled/handled by babel and made it so that now the BlueElectrum module is loaded eagerly rather than lazily. This caused some problems in the integration tests because the client was trying unexpectedly to connect to electrum; I patched the tests copying from the ones that were using BlueElectrum directly, but I'm not sure if the change in load mode might have impacted some other areas as well. Again if you think this can cause problems I can look for a way to retain the original load mode for BlueElectrum (although I have to say it doesn't look trivial...).

Overtorment commented:
id say dont rush converting everything to TS.
"f it works, don't touch it"

demerzel3 commented:
This is a way of filling the gaps that I left in the last PR as regard to `AbstractWallet`: some of the fields have underspecified types and some are just plain wrong (see `utxo` for example). Besides, I think that if we want to switch to typescript starting from the models would serve us well as an approach, so I see this as valuable work overall. It's also a great way for me to get familiar with the codebase.
Of course I don't want to break anything with this, I will make sure that all existing tests still pass and I will revert any unwanted change in behavior that I may have introduced with the types (as soon as I have a bit of time to continue working on it 😄)

demerzel3 commented:
@Overtorment I think this is ready for review now if you want to take a look, I've updated the main comment with a couple of notes.

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

limpbrains reviewed (approved):
I very much like it!

I believe we need to refactor BlueElectrum so that it will have manual connect() method, that we can run in tests if it is needed.

Overtorment commented on tests/integration/watch-only-wallet.test.js:
```diff
@@ -1,8 +1,7 @@
-global.net = require('net'); // needed by Electrum client. For RN it is proviced in shim.js
-global.tls = require('tls'); // needed by Electrum client. For RN it is proviced in shim.js
-const BlueElectrum = require('../../blue_modules/BlueElectrum'); // so it connects ASAP
+import * as BlueElectrum from '../../blue_modules/BlueElectrum';
```
there might be a problem with this approach, as `require` caches the result,
so if module that's being required creates a connection (and also some variables in its scope),  this connection is reused every time someone requires this module.
this is how we ensured.

with `import` its a different deal afaik. everything is brand new and fresh when you import a module, so sockets and connections are duplicated.

not sure how it all works in this PR, need to finish reading the diff and try running it myself

demerzel3 replied:
This is not my experience at all, as far as I know es6 modules are only evaluated once on the first import exactly like `require`, it shouldn't create any problem.

Overtorment replied:
nice

Overtorment commented on tests/integration/notifications.test.js:
```diff
@@ -1,6 +1,8 @@
+import assert from 'assert';
```
whats the purpose of changing `require` -> `import` ?

demerzel3 replied:
just consistency 🤷‍♂️

Overtorment commented on tests/integration/lightning-custodian-wallet.test.js:
```diff
@@ -1,6 +1,24 @@
+
 import { LightningCustodianWallet } from '../../class';
-const assert = require('assert');
+import * as BlueElectrum from '../../blue_modules/BlueElectrum';
```
lightning custodial wallet doesnt need `BlueElectrum`, it works through REST API

Overtorment commented on tests/setup.js:
```diff
@@ -2,6 +2,9 @@
 
 import mockClipboard from '@react-native-clipboard/clipboard/jest/clipboard-mock.js';
 
+global.net = require('net'); // needed by Electrum client. For RN it is proviced in shim.js
```
good idea to move it here

Overtorment commented on tests/integration/hodlhodl.test.js:
```diff
@@ -1,8 +1,20 @@
+
 import { LegacyWallet, SegwitBech32Wallet, SegwitP2SHWallet } from '../../class';
 import { HodlHodlApi } from '../../class/hodl-hodl-api';
+import * as BlueElectrum from '../../blue_modules/BlueElectrum';
```
hodlhodl doesnt need `BlueElectrum`

Overtorment commented on tests/integration/aopp.test.js:
```diff
@@ -2,6 +2,18 @@ import assert from 'assert';
 
 import { HDLegacyP2PKHWallet, HDSegwitP2SHWallet } from '../../class';
 import AOPP from '../../class/aopp';
+import * as BlueElectrum from '../../blue_modules/BlueElectrum'; // so it connects ASAP
```
AOPP doesnt need `BlueElectrum`

Overtorment reviewed (changes requested):
ok i see your troubles with BlueElectrum.
the easiest solution would be (I just tested it, it works) to remove mandatory connection from BlueElectrum.js (basically socket connects as soon as module is required, lets change that),
and add connection to BlueApp.js right after BlueElectrum is required.

```
overtorment@laptop3 ~/Documents/BlueWallet (master) $ git diff BlueApp.js blue_modules/BlueElectrum.js
diff --git a/BlueApp.js b/BlueApp.js
index acdce2a0..4b0a1fcf 100644
--- a/BlueApp.js
+++ b/BlueApp.js
@@ -5,6 +5,7 @@ import loc from './loc';
 const prompt = require('./blue_modules/prompt');
 const currency = require('./blue_modules/currency');
 const BlueElectrum = require('./blue_modules/BlueElectrum'); // eslint-disable-line @typescript-eslint/no-unused-vars
+BlueElectrum.connectMain();
 const BlueApp = new AppStorage();
 // If attempt reaches 10, a wipe keychain option will be provided to the user.
 let unlockAttempt = 0;
diff --git a/blue_modules/BlueElectrum.js b/blue_modules/BlueElectrum.js
index 4ff7325f..599d9a57 100644
--- a/blue_modules/BlueElectrum.js
+++ b/blue_modules/BlueElectrum.js
@@ -183,8 +183,6 @@ async function connectMain() {
   }
 }
 
-connectMain();
-
 async function presentNetworkErrorAlert(usingPeer) {
   if (await isDisabled()) {
     console.log(
```


otherwise LGTM

demerzel3 commented on class/wallets/legacy-wallet.ts:
```diff
@@ -16,13 +23,16 @@ export class LegacyWallet extends AbstractWallet {
   static type = 'legacy';
   static typeReadable = 'Legacy (P2PKH)';
 
+  _txs_by_external_index: Transaction[] = []; // eslint-disable-line camelcase
```
Not sure what you mean here, they are used in `fetchTransactions` ([here](https://github.com/demerzel3/BlueWallet/blob/82299ebaea12010d5e26d1712b14621ecf8a1890/class/wallets/legacy-wallet.ts#L346)) and `getTransactions` ([here](https://github.com/demerzel3/BlueWallet/blob/82299ebaea12010d5e26d1712b14621ecf8a1890/class/wallets/legacy-wallet.ts#L352-L354)).

demerzel3 commented on class/wallets/legacy-wallet.ts:
```diff
@@ -364,14 +411,25 @@ export class LegacyWallet extends AbstractWallet {
+    sequence: number,
+    skipSigning = false,
+    masterFingerprint: number,
+  ): CreateTransactionResult<U> {
```
here `U` is anything that extends `CreateTransactionUtxo`, it is a way of expressing that the list of utxos that you pass in must at least have the same fields as `CreateTransactonUtxo`, but whatever extra attributes you pass in it will be kept intact in the result.

If `CreateTransactionUtxo` is this type:
```ts
type CreateTransactionUtxo = {
  txId: string
  value: number
}
```
Then you can do something like this and typescript won't complain:
```ts
// given the shape of the input `U` is inferred here to be: 
// {txId: string; value: number; foo: string}
const result = createTransaction(
    [{ txId: '123', value: 123, foo: 'foo' }], 
    ... whatever ...
)
// Since in the result `inputs` is of type `U[]` we can read `foo` without issues
console.log(result.inputs[0].foo)
```

demerzel3 commented on class/wallets/legacy-wallet.ts:
```diff
@@ -392,12 +450,13 @@ export class LegacyWallet extends AbstractWallet {
+    const sanitizedOutputs = outputs.map(output => ({
+      ...output,
       // if output has no address - this is change output
-      if (!output.address) {
-        output.address = changeAddress;
-      }
+      address: output.address ?? changeAddress,
+    }));
```
the existing code was fine but this is better because it changes the type of `address` from being optional (`string | undefined`) to always being present (`string`), which is a tad safer and saves some checks down the line.

demerzel3 commented on typings/coinselect.d.ts:
```diff
@@ -0,0 +1,43 @@
+    txId: string;
+  };
+
+  export default function coinSelectSplit<U extends Utxo>(
```
Same as a previous comment, `U` gets inferred from the input to be anything that extends the `Utxo` type; whatever extra attribute the input has will be preserved in the result (as you can see `inputs` is of type `U[]` in the return type).

demerzel3 commented:
@Overtorment I moved `BlueElectrum` connection to `BlueApp` as per your suggestion and reverted the changes to the tests that didn't need `BlueElectrum` in the first place.

Overtorment commented on tests/integration/hd-legacy-breadwallet.test.js:
```diff
@@ -17,7 +16,7 @@ afterAll(async () => {
   // awaiting for Electrum to be connected. For RN Electrum would naturally connect
   // while app starts up, but for tests we need to wait for it
-  await BlueElectrum.waitTillConnected();
+  await BlueElectrum.connectMain();
```
you sure we dont need both `connectMain` & `waitTillConnected` ?

Overtorment commented on tests/integration/import.test.js:
```diff
@@ -58,7 +57,7 @@ afterAll(async () => {
 beforeAll(async () => {
   // awaiting for Electrum to be connected. For RN Electrum would naturally connect
```
not precisely correct comment anymore

Overtorment commented:
running all tests here: https://github.com/BlueWallet/BlueWallet/pull/3576
